core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java

```diff
@@ -426,6 +426,10 @@ public void append(Key key, Value value) throws IOException {
         throw new IllegalArgumentException("Keys appended out-of-order.  New key " + key + ", previous key " + prevKey);
       }
 
+      if (((long) key.getSize() + (long) value.getSize()) >= Integer.MAX_VALUE) {
```

Comment:
Serialization will also add overhead. For example some bytes encoding the length of the row are written before the rows data.  So this means that what can really be written is less than what is checked for here.  Not sure if we want to try to account for that here.  We could do something simple like assume each field will need at least 5 bytes for the length (I think 5 is the max because we are using var int).  

Even if not accounting for that serialization overhead this is still a really nice check.  It will catch most situations at the source leaving a small range of key+val size that would only be caught later when encoding the block.